src/tls.rs

```diff
@@ -1,6 +1,354 @@
+        }
+        let ret = unsafe {
+            esp_idf_sys::esp_tls_conn_new_sync(
+                host.as_bytes().as_ptr() as *const i8,
+                host.len() as i32,
+                port as i32,
+                &rcfg,
+                tls,
+            )
+        };
+
+        if ret == 1 {
+            Ok(EspTls {
+                reader: EspTlsRead { raw: tls },
+                writer: EspTlsWrite { raw: tls },
+            })
+        } else {
+            unsafe {
+                esp_idf_sys::esp_tls_conn_destroy(tls);
+            }
+
+            Err(EspError::from_infallible::<ESP_FAIL>())
+        }
+    }
+
+    pub fn read(&mut self, buf: &mut [u8]) -> Result<usize, EspIOError> {
+        self.reader.read(buf)
+    }
+
+    pub fn write(&mut self, buf: &[u8]) -> Result<usize, EspIOError> {
+        self.writer.write(buf)
+    }
+
+    pub fn split(&mut self) -> (&mut EspTlsRead, &mut EspTlsWrite) {
+        (&mut self.reader, &mut self.writer)
+    }
+}
+
+impl io::Io for EspTls {
+    type Error = EspIOError;
+}
+
+impl io::Read for EspTls {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, EspIOError> {
+        self.read(buf)
+    }
+}
+
+impl io::Write for EspTls {
+    fn write(&mut self, buf: &[u8]) -> Result<usize, EspIOError> {
+        self.write(buf)
+    }
+
+    fn flush(&mut self) -> Result<(), EspIOError> {
+        Ok(())
+    }
+}
+
+pub struct EspTlsRead {
+    raw: *mut esp_idf_sys::esp_tls,
+}
+
+impl EspTlsRead {
+    pub fn read(&mut self, buf: &mut [u8]) -> Result<usize, EspIOError> {
+        // cannot call esp_tls_conn_read bc it's inline
+        //let ret = unsafe { esp_idf_sys::esp_tls_conn_read(self.raw, buf.as_mut_ptr(), buf.len()) };
+        let esp_tls = unsafe { std::ptr::read_unaligned(self.raw) };
+        let read_func = esp_tls.read.unwrap();
+        let ret = unsafe { read_func(self.raw, buf.as_mut_ptr() as *mut i8, buf.len()) };
+        // ESP docs treat 0 as error, but in Rust it's common to return 0 from `Read::read` to indicate eof
+        if ret >= 0 {
+            Ok(ret as usize)
+        } else {
+            Err(EspIOError(EspError::from(ret as i32).unwrap()))
+        }
+    }
+}
+
+impl io::Io for EspTlsRead {
+    type Error = EspIOError;
+}
+
+impl io::Read for EspTlsRead {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, EspIOError> {
+        self.read(buf)
+    }
+}
+
+pub struct EspTlsWrite {
+    raw: *mut esp_idf_sys::esp_tls,
+}
+
+impl EspTlsWrite {
+    pub fn write(&mut self, buf: &[u8]) -> Result<usize, EspIOError> {
+        // cannot call esp_tls_conn_write bc it's inline
+        //let ret = unsafe { esp_idf_sys::esp_tls_conn_write(self.raw, buf.as_ptr(), buf.len()) };
+        let esp_tls = unsafe { std::ptr::read_unaligned(self.raw) };
+        let write_func = esp_tls.write.unwrap();
+        let ret = unsafe { write_func(self.raw, buf.as_ptr() as *const i8, buf.len()) };
```

Comment:
Ditto, if ESP-TLS treats writing a 0-length buffer as an error. That's OK in Rust, and in that case we should return `Ok(0)`, as that's how the `Read`/`Write` traits contract says.